4.x: TestSpanListenersWithInjection failing in validation workflow

This test has started failing in the validation workflow. Maybe a poor interaction between test added by PR #9718 and MP testing rework in #9695.


```
[INFO] Running io.helidon.microprofile.telemetry.TestSpanListenersWithInjection
Error:  Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 1.264 s <<< FAILURE! - in io.helidon.microprofile.telemetry.TestSpanListenersWithInjection
Error:  io.helidon.microprofile.telemetry.TestSpanListenersWithInjection.testSpanFromInjectedTracer  Time elapsed: 1.023 s  <<< ERROR!
java.lang.RuntimeException: 
java.lang.AssertionError: After span start: Starting spans
Expected: a collection with size <1>
     but: collection size was <0>
	at io.helidon.microprofile.testing.ReflectionHelper.invoke(ReflectionHelper.java:295)
	at io.helidon.microprofile.testing.Instrumented.lambda$instrument$0(Instrumented.java:103)
	at io.helidon.microprofile.telemetry.TestSpanListenersWithInjection$ByteBuddy$Z2LCPArV.testSpanFromInjectedTracer(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at io.helidon.testing.junit5.TestJunitExtension.lambda$invoke$8(TestJunitExtension.java:373)
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:137)
	at io.helidon.testing.junit5.TestJunitExtension.invoke(TestJunitExtension.java:371)
	at io.helidon.testing.junit5.TestJunitExtension.interceptTestMethod(TestJunitExtension.java:144)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
```

--- a/.github/actions/common/action.yml
+++ b/.github/actions/common/action.yml
@@ -84,7 +84,7 @@ runs:
         java-version: ${{ env.JAVA_VERSION }}
     - name: Cache local Maven repository (read-write)
       if: ${{ inputs.maven-cache == 'read-write' }}
-      uses: actions/cache@v4.0.2
+      uses: actions/cache@v4.2.0
       with:
         # See https://github.com/actions/toolkit/issues/713
         # Include must not match top level directories
@@ -98,7 +98,7 @@ runs:
           local-maven-
     - name: Cache local Maven repository (read-only)
       if: ${{ inputs.maven-cache == 'read-only' }}
-      uses: actions/cache/restore@v4.0.2
+      uses: actions/cache/restore@v4.2.0
       with:
         path: |
           .m2/repository/**/*.*
@@ -109,7 +109,7 @@ runs:
           local-maven-
     - name: Build cache (read-write)
       if: ${{ inputs.build-cache == 'read-write' }}
-      uses: actions/cache@v4.0.2
+      uses: actions/cache@v4.2.0
       with:
         path: |
           ./**/target/**
@@ -121,7 +121,7 @@ runs:
           build-cache-${{ github.run_id }}-
     - name: Build cache (read-only)
       if: ${{ inputs.build-cache == 'read-only' }}
-      uses: actions/cache/restore@v4.0.2
+      uses: actions/cache/restore@v4.2.0
       with:
         path: |
           ./**/target/**

--- a/common/common/src/main/java/io/helidon/common/Functions.java
+++ b/common/common/src/main/java/io/helidon/common/Functions.java
@@ -246,21 +246,4 @@ public final class Functions {
          */
         T get() throws E;
     }
-
-    /**
-     * Unchecked exception.
-     *
-     * @see #unwrap(Throwable)
-     */
-    public static class UncheckedException extends RuntimeException {
-
-        /**
-         * Create a new unchecked exception.
-         *
-         * @param cause cause
-         */
-        public UncheckedException(Throwable cause) {
-            super(cause);
-        }
-    }
 }

--- a/common/common/src/main/java/io/helidon/common/UncheckedException.java
+++ b/common/common/src/main/java/io/helidon/common/UncheckedException.java
@@ -1,0 +1,31 @@
+/*
+ * Copyright (c) 2025 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.common;
+
+/**
+ * Unchecked exception.
+ */
+public class UncheckedException extends RuntimeException {
+
+    /**
+     * Create a new unchecked exception.
+     *
+     * @param cause cause
+     */
+    public UncheckedException(Throwable cause) {
+        super(cause);
+    }
+}

--- a/microprofile/telemetry/src/test/java/io/helidon/microprofile/telemetry/TestSpanListenersWithInjection.java
+++ b/microprofile/telemetry/src/test/java/io/helidon/microprofile/telemetry/TestSpanListenersWithInjection.java
@@ -28,8 +28,10 @@ import io.opentelemetry.api.trace.Span;
 import io.opentelemetry.api.trace.SpanBuilder;
 import io.opentelemetry.api.trace.Tracer;
 import io.opentelemetry.context.Scope;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.context.Initialized;
+import jakarta.enterprise.event.Observes;
 import jakarta.inject.Inject;
-import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
@@ -41,21 +43,18 @@ import static org.hamcrest.Matchers.hasSize;
 @AddBean(TestSpanListenersWithInjection.MyBean.class)
 class TestSpanListenersWithInjection {
 
-    private static MyListener listener;
+    private static final MyListener LISTENER = new MyListener();
 
     @Inject
     private MyBean myBean;
 
-    @BeforeAll
-    static void setup() {
-        listener = new MyListener();
-        io.helidon.tracing.Tracer.global().register(listener);
-    }
-
     @BeforeEach
     void clearListener() {
-        listener.clear();
+        LISTENER.clear();
+    }
 
+    void onStartup(@Observes @Initialized(ApplicationScoped.class) Object event) {
+        io.helidon.tracing.Tracer.global().register(LISTENER);
     }
 
     @Test
@@ -108,13 +107,14 @@ class TestSpanListenersWithInjection {
                                      int expectedEnded,
                                      int expectedActivated,
                                      int expectedClosed) {
-        assertThat(message + ": Starting spans", listener.starting(), hasSize(expectedStarting));
-        assertThat(message + ": Started spans", listener.started(), hasSize(expectedStarted));
-        assertThat(message + ": Ended spans", listener.ended(), hasSize(expectedEnded));
-        assertThat(message + ": Activated scopes", listener.activated().keySet(), hasSize(expectedActivated));
-        assertThat(message + ": Closed scopes", listener.closed().keySet(), hasSize(expectedClosed));
+        assertThat(message + ": Starting spans", LISTENER.starting(), hasSize(expectedStarting));
+        assertThat(message + ": Started spans", LISTENER.started(), hasSize(expectedStarted));
+        assertThat(message + ": Ended spans", LISTENER.ended(), hasSize(expectedEnded));
+        assertThat(message + ": Activated scopes", LISTENER.activated().keySet(), hasSize(expectedActivated));
+        assertThat(message + ": Closed scopes", LISTENER.closed().keySet(), hasSize(expectedClosed));
     }
 
+    @ApplicationScoped
     static class MyBean {
 
         @Inject

--- a/microprofile/testing/testing/src/main/java/io/helidon/microprofile/testing/Instrumented.java
+++ b/microprofile/testing/testing/src/main/java/io/helidon/microprofile/testing/Instrumented.java
@@ -19,11 +19,14 @@ import java.lang.annotation.Annotation;
 import java.lang.invoke.MethodHandles;
 import java.lang.reflect.Field;
 import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.Arrays;
 import java.util.List;
 import java.util.function.BiFunction;
 import java.util.stream.Stream;
+
+import io.helidon.common.UncheckedException;
 
 import net.bytebuddy.ByteBuddy;
 import net.bytebuddy.description.method.MethodDescription;
@@ -100,7 +103,14 @@ public interface Instrumented {
         return instrument(type, annotations, methodExcludes, (proxy, method, args) -> {
             T instance = resolver.apply(type, method);
             if (instance != null) {
-                return invoke(Object.class, method, instance, args);
+                try {
+                    return invoke(Object.class, method, instance, args);
+                } catch (UncheckedException e) {
+                    if (e.getCause() instanceof InvocationTargetException te) {
+                        throw te;
+                    }
+                    throw e;
+                }
             }
             return null;
         });

--- a/microprofile/testing/testing/src/main/java/io/helidon/microprofile/testing/ReflectionHelper.java
+++ b/microprofile/testing/testing/src/main/java/io/helidon/microprofile/testing/ReflectionHelper.java
@@ -30,6 +30,8 @@ import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+import io.helidon.common.UncheckedException;
+
 /**
  * Reflection helper.
  */
@@ -285,14 +287,8 @@ class ReflectionHelper {
             method.setAccessible(true);
             Object value = method.invoke(instance, args);
             return type.cast(value);
-        } catch (IllegalAccessException e) {
-            throw new RuntimeException(e);
-        } catch (InvocationTargetException e) {
-            Throwable target = e.getTargetException();
-            if (e.getTargetException() instanceof RuntimeException re) {
-                throw re;
-            }
-            throw new RuntimeException(target);
+        } catch (IllegalAccessException | InvocationTargetException e) {
+            throw new UncheckedException(e);
         }
     }
 }

--- a/microprofile/testing/testng/src/main/java/io/helidon/microprofile/testing/testng/HelidonTestNgListener.java
+++ b/microprofile/testing/testng/src/main/java/io/helidon/microprofile/testing/testng/HelidonTestNgListener.java
@@ -31,7 +31,7 @@ import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Semaphore;
 
-import io.helidon.common.Functions.UncheckedException;
+import io.helidon.common.UncheckedException;
 import io.helidon.common.context.Context;
 import io.helidon.common.context.Contexts;
 import io.helidon.microprofile.testing.HelidonTestContainer;
@@ -268,13 +268,13 @@ public class HelidonTestNgListener extends HelidonTestNgListenerBase implements 
                     method.setAccessible(true);
                     return method.invoke(instance, args);
                 } catch (InvocationTargetException e) {
-                    if (e.getTargetException() instanceof RuntimeException) {
-                        throw (RuntimeException) e.getTargetException();
-                    }
-                    throw new UncheckedException(e.getTargetException());
+                    throw new UncheckedException(e);
                 }
             });
         } catch (UncheckedException e) {
+            if (e.getCause() instanceof InvocationTargetException te) {
+                throw te.getCause();
+            }
             throw e.getCause();
         }
     }

